Sources/apollo-ios-pagination/GraphQLQueryPager.swift

```diff
@@ -0,0 +1,204 @@
+        guard let self else { return }
+        switch result {
+        case .success(let data):
+          if case .server = data.source, let firstPageData = data.data {
+            let page = Hashed(wrappedValue: self.extractPageInfo(.initial(firstPageData)))
+            pageOrder.append(page)
+          }
+          self.initialPageResult = data.data
+          if let latest = self.latest {
+            let (firstPage, nextPage) = latest
+            self.onUpdate?((firstPage, nextPage, data.source == .cache ? .cache : .fetch))
+          }
+        case .failure(let error):
+          self.onError?(error)
+        }
+      }
+    )
+  }
+
+  /// Loads the next page, based on the latest page info.
+  public func loadMore(
+    cachePolicy: CachePolicy = .fetchIgnoringCacheData,
+    completion: (() -> Void)? = nil
+  ) async throws {
+    guard let currentPageInfo else {
+      assertionFailure("No page info detected -- are you calling `loadMore` prior to calling the initial fetch?")
+      throw PaginationError.missingInitialPage
+    }
+    guard let nextPageQuery = nextPageResolver(currentPageInfo.wrappedValue),
+          currentPageInfo.wrappedValue.canLoadMore
+    else { throw PaginationError.pageHasNoMoreContent }
+    guard activeTask == nil else {
+      _ = await activeTask?.value
+      return
+    }
+    let task = Task<Void, Never> {
+      let watcher = GraphQLQueryWatcher(client: client, query: nextPageQuery) { [weak self] result in
+        defer { self?.activeTask = nil }
+        guard let self else { return }
+        switch result {
+        case .success(let data):
+          guard let nextPageData = data.data else { return }
+
+          let page = Hashed(wrappedValue: self.extractPageInfo(.paginated(nextPageData)))
+          if case .server = data.source {
+            pageOrder.append(page)
+          }
+          self.pageMap[page] = data.data
+
+          if let latest = self.latest {
+            let (firstPage, nextPage) = latest
+            self.onUpdate?((firstPage, nextPage, data.source == .cache ? .cache : .fetch))
+          }
+        case .failure(let error):
+          self.onError?(error)
+        }
+        completion?()
+      }
+      nextPageWatchers.append(watcher)
+      watcher.refetch(cachePolicy: cachePolicy)
+    }
+
+    await task.value
+  }
+
+  /// Reloads all data, starting at the first query, resetting pagination state.
+  public func refetch() {
+    cancel()
+    fetch()
+  }
+
+  /// Cancel any in progress fetching operations and unsubscribe from the store.
+  public func cancel() {
+    nextPageWatchers.forEach { $0.cancel() }
+    nextPageWatchers = []
+    firstPageWatcher?.cancel()
+    firstPageWatcher = nil
+
+    pageMap = [:]
+    pageOrder = []
+    initialPageResult = nil
+  }
+
+  deinit {
+    firstPageWatcher?.cancel()
+    nextPageWatchers.forEach { $0.cancel() }
+  }
+}
+
+@propertyWrapper
+private struct Hashed<Wrapped>: Hashable {
```

Comment:
Really just a hack to not have to include a `PaginationInfo` as a generic type argument. I want all `PaginationInfo` to be `Hashable`, but I don't want to add that as an explicit requirement because it complicates usage.

Instead, I'm wrapping here and calling an `assertionFailure` in the case that the passed `PaginationInfo` isn't `Hashable`. Could elevate the `assertionFailure` into a `preconditionFailure`, but I'm wary of doing so.